src/IssueViz/Editor/QuickActions/IssueLocationActionsSource.cs

```diff
@@ -0,0 +1,123 @@
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Tagging;
+using SonarLint.VisualStudio.IssueVisualization.Editor.LocationTagging;
+using SonarLint.VisualStudio.IssueVisualization.Editor.SelectedIssueTagging;
+using SonarLint.VisualStudio.IssueVisualization.Models;
+using SonarLint.VisualStudio.IssueVisualization.Selection;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Editor.QuickActions
+{
+    internal sealed class IssueLocationActionsSource : ISuggestedActionsSource
+    {
+        private readonly IAnalysisIssueSelectionService selectionService;
+        private readonly ITagAggregator<IIssueLocationTag> issueLocationsTagAggregator;
+        private readonly ITagAggregator<ISelectedIssueLocationTag> selectedIssueLocationsTagAggregator;
+
+        public IssueLocationActionsSource(IBufferTagAggregatorFactoryService bufferTagAggregatorFactoryService, ITextBuffer textBuffer, IAnalysisIssueSelectionService selectionService)
+        {
+            this.selectionService = selectionService;
+
+            issueLocationsTagAggregator = bufferTagAggregatorFactoryService.CreateTagAggregator<IIssueLocationTag>(textBuffer);
+            issueLocationsTagAggregator.TagsChanged += TagAggregator_TagsChanged;
+
+            selectedIssueLocationsTagAggregator = bufferTagAggregatorFactoryService.CreateTagAggregator<ISelectedIssueLocationTag>(textBuffer);
+            selectedIssueLocationsTagAggregator.TagsChanged += TagAggregator_TagsChanged;
+        }
+
+        public event EventHandler<EventArgs> SuggestedActionsChanged;
+
+        public IEnumerable<SuggestedActionSet> GetSuggestedActions(ISuggestedActionCategorySet requestedActionCategories, SnapshotSpan range, CancellationToken cancellationToken)
+        {
+            var allActions = new List<ISuggestedAction>();
+
+            if (IsOnIssueVisualization(range, out var issueVisualizations))
+            {
+                var actions = issueVisualizations.Select(x => new SelectIssueVisualizationAction(selectionService, x));
+                allActions.AddRange(actions);
+            } 
+
+            if (IsOnSelectedVisualization(range))
+            {
+                allActions.Add(new DeselectIssueVisualizationAction(selectionService));
+            }
+
+            if (allActions.Any())
+            {
+                return new[] {new SuggestedActionSet(allActions)};
+            }
+
+            return Enumerable.Empty<SuggestedActionSet>();
+        }
+
+        public Task<bool> HasSuggestedActionsAsync(ISuggestedActionCategorySet requestedActionCategories, SnapshotSpan range, CancellationToken cancellationToken)
+        {
+            return Task.Factory.StartNew(() => IsOnIssueVisualization(range, out _) || IsOnSelectedVisualization(range),
+                cancellationToken);
+        }
+
+        private bool IsOnIssueVisualization(SnapshotSpan range, out IEnumerable<IAnalysisIssueVisualization> issueVisualizationsWithSecondaryLocations)
```

Comment:
The method name is bit misleading.
`IsOnIssueWithSecondaryLocations` / `ContainsIssuesWithSecondaryLocations` would be clearer.